docker/Dockerfile.leap.42.3

```diff
@@ -0,0 +1,59 @@
+#
+# Copyright 2019, Intel Corporation
+#
+# 'recipe' for Docker to build an RPM
+#
+
+# Pull base image
+FROM opensuse/leap:42.3
+MAINTAINER daos-stack <daos@daos.groups.io>
+
+# use same UID as host and default value of 1000 if not specified
+ARG UID=1000
+
+# Add build user (to keep rpmbuild happy)
+ENV USER build
+RUN useradd -u $UID -ms /bin/bash $USER
+RUN groupadd -g $UID $USER
+
+# Install basic tools
+RUN zypper --non-interactive update
+
+# basic building
+ENV OPENSUSE_REPO="https://download.opensuse.org/repositories"
+ENV TOOLS_42_3A="openSUSE:/Tools/openSUSE_42.3/openSUSE:Tools.repo"
+ENV TOOLS_42_3="${OPENSUSE_REPO}/${TOOLS_42_3A}"
+ENV HPC_42_3A="science:/HPC/openSUSE_Leap_42.3/science:HPC.repo"
+ENV HPC_42_3="${OPENSUSE_REPO}/${HPC_42_3A}"
+
+RUN zypper --non-interactive ar -f ${TOOLS_42_3}; \
+    zypper --non-interactive ar -f ${HPC_42_3}; \
+    zypper --non-interactive --gpg-auto-import-keys refresh
+RUN zypper --non-interactive install autoconf automake build                  \
+                                     ca-certificates-mozilla curl createrepo  \
+                                     git libtool lsb-release make             \
+                                     perl-libwww-perl perl-LWP-Protocol-https \
+                                     perl-XML-Parser rpm-build sudo
+
+# need to run the build command as root, as it needs to chroot
+RUN echo "build ALL=(ALL) NOPASSWD: /usr/bin/build" > /etc/sudoers.d/build
+
+# Need the repo for munge
+ARG JENKINS_URL=""
+ENV JENKINS_PATH="${JENKINS_URL}job/daos-stack/job"
+ENV JENKINS_ART="lastSuccessfulBuild/artifact/artifacts"
+ARG MNGE_BRANCH="master"
+RUN zypper --non-interactive ar --gpgcheck-allow-unsigned -f \
+    ${JENKINS_PATH}/munge/job/${MNGE_BRANCH}/${JENKINS_ART}/leap42.3 munge; \
+    zypper --non-interactive ref munge
+
+
+# Suse build utility can not handle spaces in the repo aliases.
+RUN zypper --non-interactive nr 'NON OSS' NON_OSS; \
+    zypper --non-interactive nr 'NON OSS Update' NON_OSS_Update; \
+    zypper --non-interactive nr 'OSS Update' OSS_Update
+
+# Suse build utility needs this refresh.
```

Comment:
Why does it need the refresh?  The repos it's using should be in the chroot and updated within it.

Reply:
That may be a bug in the build utility.  A Leap Docker file from dockerhub has repository already comes with data in it that build considers to be expired so build will not use it.  Same with our locally built SLES docker containers.
Zypper is not bothered by this data being expired but the Suse build utility is, so this is normally not noticed.

If we do not have this step at the end of the Dockerfile with a cachebuster, eventually the Suse builds will start failing from expired repository data.

Reply:
> That may be a bug in the build utility. A Leap Docker file from dockerhub has repository already comes with data in it that build considers to be expired so build will not use it.

All the more reason to add the repo to the `build` command and let it fetch the repo data fresh.

> If we do not have this step at the end of the Dockerfile with a cachebuster, eventually the Suse builds will start failing from expired repository data.

Not if we add the repos at the `build` command level where the repo data is fetch fresh for every build.

Reply:
Again, this looks like more problems with just not having the `build` tool add the repos in the chroot.

We seem to be doing a lot of work-arounds just to avoid using `--repo ...` `build` arguments.

Reply:
The distribution repos for SLES/Leap will only work with the --repo zypp:// option because they do not have a repodata directory.  Adding the repodata directory to them breaks zypper access to them.
And those are the repositories that also need to be refreshed.

So unless we get the Suse build utility changed to work like zypper, there is no way around either the --repo zypp:// or forcing a refresh of the repositories one way or another.

For CentOS, the repositories are passed to the chroot via a data file on the container, so using the zypp:// option makes the SLES/Leap work more like the CentOS.

And if we do not use the --repo zypp:// option, then we either have to add a bunch of parameters that are almost the same, but specific to each distro/version either to every jenkinsfile, or add the generation of a table for Suse to the Makefile_packaging or makefile.  Both of which appear to be options that will harder to maintain.

Reply:
> The distribution repos for SLES/Leap will only work with the --repo zypp:// option because they do not have a repodata directory.

https://github.com/openSUSE/zypper/wiki/man_page indicates that `addrepo` has a `-t`/`--type` option with several different types of repos.  One is `yast2` and yet another is `plaindir` and another which says _Add a HTTP repository, probe it..._.

I wonder if any of these types makes the SLES12/Leap repos usable without `repodata`.

> For CentOS, the repositories are passed to the chroot via a data file on the container, so using the zypp:// option makes the SLES/Leap work more like the CentOS.

The key difference is that it's just a list of repositories to add to the chroot where the repodata for those is fetched at package build time.  The problem with adding the repos to the container is that the container's repodata is used by `build` which gets stale and the container needs refreshing.  The CentOS container doesn't need refreshing because of any similar problems.  I would actually compare the big repo file that mock uses more like the `--repo` arguments to `build` as those both result in the repos being built inside the chroot with fresh repodata.

> And if we do not use the --repo zypp:// option, then we either have to add a bunch of parameters that are almost the same, but specific to each distro/version either to every jenkinsfile, or add the generation of a table for Suse to the Makefile_packaging or makefile. Both of which appear to be options that will harder to maintain.

I think maintaining the list of repos for each distro all in one place -- the `Makefile_packaging.mk` is actually easier than having them spread across all of the Dockerfiles.

But ultimately, the real goal here is the freshest metadata for the chroot-build tool without causing docker image churn on every build.  The chroots are ephemeral and only last as long as the job.  The docker images are much longer lived and will pile up in the cache, which has always been a concern of yours.